### PR TITLE
New dashboard grid size + scalable cards + fullscreen + auto refresh

### LINKED ISSUE #913: Pie Chart Legends shouldn't overlap the pie chart
See attached.
screen shot 2015-08-26 at 12 51 29 pm

### LINKED ISSUE #1024: Order of dashboard cards on a small screen should be same order as on a large screen
Presently, the order of cards on a customized dashboard when viewed on a large screen are one way, and then the order changes in an unexpected way when you resize your window down to tablet/mobile size. Instead of presenting the cards stacked vertically in the expected order (i.e., top-left card first, the one to the right of it second, and so on), the cards are presented *in the order in which they were created*. This is unexpected.

Cards arranged how I want them. **Card E was created first.**
screen shot 2015-09-23 at 2 18 27 pm

Cards all willy nilly:
screen shot 2015-09-23 at 2 18 46 pm

### LINKED ISSUE #1166: Pin maps should use pins that don't look like zits
Guessing we could use Google Maps' built-in pin icons.

screen shot 2015-10-07 at 11 48 54 am

### LINKED ISSUE #1349: Dashboard fullscreen / auto-refresh ability
Currently the width is limited and the widgets are centered, which doesn't allow to take advantage of bigger resolutions.

It could be nice to have an option to display the dashboard full screen and also auto-refreshing the metrics, typically to set up a dashboard on a TV. 
 
(from http://discourse.metabase.com/t/dashboard-on-tv/51)

### LINKED ISSUE #2055: Lots of empty space at the bottom of my dashboard
If this is a bug report, please fill in the blanks:

*  I am using the OSX app
*  My Metabase version is `0.15.0`

Here's a screenshot of `0.14.1` (left) side-by-side with `0.15.0` (right), both scrolled all the way to the bottom of the same dashboard. `0.14.1` stops where the last card stops but `0.15.0` lets me keep scrolling well past the end of the board.

screen shot 2016-02-29 at 3 51 37 pm

FWIW I've tried this in the `0.12.1` and `0.13.3` releases of the OS X app as well. Those behave like `0.14.1` (no extra whitespace)

### LINKED ISSUE #2076: Pin Map keeps zooming on San Francisco
I have a pin map added to my dashboard. By default the pin map shows the San Francisco area. Even if I change it to display the world, it resets to SF. Is there a way to make it display the entire world by default.

*  I am using the Chrome browser.
*  My computer's OS is Mac OS
*  My database is Postgres
*  My Metabase version is v0.15.0 built on 2016-02-29

### LINKED ISSUE #2085: Visualization overwrites column ordering with A-Z order
If I arrange table rows for visualization in a particular order before v0.15.0 the order was kept when I switched from table to a visualization. Now visualized they are re-ordered:
image
is visualized as: 
image

### LINKED ISSUE #2094: Bar-, Line-, Area charts: labels & colors for multiple series
First of all: Thanks for the new data labels in v0.15.0. They are awesome.

One problem I found with them is with stacked bar charts. Take this table:
image

The label will show "1st value" on all parts of the bar. It should take the column header as the label. The value in the label is correct.
image
This also happens with area and line charts:
image
image

And that last screenshot shows nicely the second problem: For Barcharts and Linecharts every Line after the 1st are yellow. Especially for Barcharts that makes different series indistinguishable. Area charts correctly pick a different color for each column. 

A third problem for me is that line charts shouldn't be "stacked" like the other two. But I can work around that by using the new multiseries charts on dashboards.

### LINKED ISSUE #2119: Stacked bar charts show the same label for all series on the tooltip
If this is a bug report, please fill in the blanks:

*  I am using the Chrome browser.
*  My computer's OS is Mac.
*  My database is all.
*  My Metabase version is v0.15.0

The y-axis label should change as you hover each different series in the stacked bar, but currently it doesn't and you always see the same label.  The value itself does change though, so this is specifically about the label not changing.

### LINKED ISSUE #2125: The scalar visualization-type should scale the font-size properly
Whe using the number card-type and resizing the card on a Dashboard the inner value should fill the container. I attached ugly examples, but useful (dummy data).

today we have:
bugreport

I think the better idea is:
bugreport - copia

### LINKED ISSUE #2126: We need a "resize" icon on the corner of cards
We can resize the card size on a dashboard by draging the bottom right corner, but there is no visual clue that we can do that. There should be an icon or something that guides the user to use this feature.

### LINKED ISSUE #2141: Empty row causes pie chart to be incorrect.
I am using the mac standalone app.
My computer's OS is OSX El Capitan.
My database is Mysql 5.6.1.
My Metabase version is 0.15.1.1.

screen shot 2016-03-11 at 13 00 59
screen shot 2016-03-11 at 13 01 10

### LINKED ISSUE #2145: Tooltip on dashboard showing up on a different card
Some strange condition where I hover on card A, then go over to card B and hover but the tooltip continues to show on card A.

screen shot 2016-03-11 at 2 19 10 pm

If this is a bug report, please fill in the blanks:

*  I am using the Chrome browser.
*  My computer's OS is OSX Yosemite.
*  My database is postgres.
*  My Metabase version is 0.16.0-snapshot.

### PR DESCRIPTION
**Pull request by tlrobinson**
resolves #913 (pie chart legends overlapping)
resolves #1024 (mobile dashcard layout should be rougly same order as grid)
resolves #1166 (pin map zits)
resolves #1349 (fullscreen / auto-refresh)
resolves #2055 (empty space at bottom of dashboard)
resolves #2076 (save pin map zoom/pan)
resolves #2085 (ordinal axis original order) 
resolves #2094 (stacked chart colors + tooltips + remove line stacking)
resolves #2119 (stacked chart tooltips)
resolves #2125 (scalar scaling)
resolves #2126 (resize indicator on dash cards)
resolves #2141 (pie chart with empty values)
resolves #2145 (tooltip wonkiness)

 todo/bugs:
- [x] fix line/area/bar tooltips
- ~~fix dashboard width to work with new root layout changes~~
- ~~fullscreen dashboard sometimes doesn't fill screen (need repro steps from @mazameli)~~
- [x] night mode transition?
- ~~dashboard button dividers?~~
- [x] pie chart small "other" slice
- [x] more scalable table cards
- ~~formatting code attempting to parse values into datetimes (seeing this on pie chart, tooltip values, etc)~~
- ~~on the smallest card size (2x2) i'm only seeing 1 character of the card title with an ellipses. e.g "C..." even though there is a reasonable amount of extra space available.~~
- [x] if you start auto-refresh then go into edit mode you can add cards that get removed when the timer goes off.  we should probably kill the auto-refresh timer when the user enters edit mode of any kind.
- [x] we probably don't need to display tooltips on hover for small cards when they have non truncated values like "32" for example (the logic to show 3,301 for a tooltip as 3.3k is basically applying to all values at smaller sizes i think)
- [x] i found a strange instance where I was able to paginate beyond the total number of rows
- [x] it seems like when auto refresh is on it flashes the favicon?
- [x] the card defaults are definitely a bit too small. i'm thinking we want an initial size of at least 4x4
- [x] there seems to be a bit of a delay between flipping the color on the active sorted column in tables
- [x] the column sort icon should show when you hover on the column heading to indicate they are sortable (like the table on the query builder page) 
- [x] investigate memory leak when leaving autorefresh on
- [x] calculation for row height on dash tables is a little off (maybe just when there is paging?).  seeing a row of data show up below the paging controls.
- [x] fix dashboards on mobile
- [ ] issue with donut slice showing wrong info in center ring on hover (tooltip is correct, other slices okay)

style:
- [x] ◀ and ▶ icons (@kdoh)
- [x] padding/margin on card grid doesn't align w/ the icon buttons in the header.  extra p2 on the right?
- [x] there's an extra pr2 on the right side of the dashboard
- [x] dots in night mode are white instead of line color
- [x] empty dashboard add button doesn't look right new fullscreen/refresh buttons
- [x] pointer cursor on click targets (dash header icons, etc)
- [x] the icons on the dash header need a bit more horizontal spacing (i think @mazameli alluded to this) 
- [x] we should have cursor pointer on table row pagination on hover
- [x] the icons for the table row pagination icons aren't properly lined up
- [x] sun and moon icons for full screen should be switched ( i should see the sun in night mode to indicate i can switch and vice versa)
- [x] when you add a new card the yellow highlight fades out over time and the card background is transparent.
- [x] the horizontal table scrollbar should only show on hover
- [x] some white padding is showing up on the edges of the light blue editing header bar on dashboards.
- [x] fix big dots in Safari
- [x] oversized card titles in QB (note: only seems to happen when there is no description)
- [x] undersized scalars in QB
- [x] extra space consumed by legend on donuts for wide cards and in QB

### CODE REVIEW

File: frontend/src/visualizations/Scalar.css [hunk 1]
```diff
@@ -0,0 +1,31 @@
+:local .Scalar {
+  display: flex;
```
Comment (kdoh) [inline frontend/src/visualizations/Scalar.css]: most of this declaration could be subsumed by using the atomic classes we have. really the only thing that seems necessary to specify here is the color which we might want to turn into a value in colors.css

**Comment by mazameli** [comment]
Looking good, man. Here's a list of things I noticed on our stats instance:


I sometimes experienced a bug in Chrome where the dash wouldn't correctly scale when I went to full screen (note the huge empty white gap on the right and bottom):
screen shot 2016-03-14 at 1 12 19 pm
When I hit esc to leave full Chrome's screen mode, Metabase kept the dashboard in the fullscreen style, even though the window was no longer full-screened:
screen shot 2016-03-14 at 1 12 37 pm

Strange bug with tooltips. When hovering over the line chart to try to see the tooltips, they would weirdly show up on the chart above, like in this screenshot:
tooltip-bug

There's also a small funny gap at the top of the pie chart:
screen shot 2016-03-14 at 11 52 47 am

Many click targets that should change the cursor to a pointer are failing to do so:
- the fullscreen icon
- all of the options in the Auto Refresh dropdown
- In fullscreen, the night mode icon and the back-to-regular-size icon
- the pie chart segments  (I think they should go to a pointer because the bar charts do, and because they invoke a tooltip — either that or bar chart segments shouldn't become a hand pointer on hover)
- scalars (same reasoning as with pie chart segments)
- points on line charts

How hard would it be to have a CSS transition when toggling between day and night modes? The transition right now is abrupt, and it would be nice polish to have it fade from one style to the other, with like a 1 second transition.

@kdoh: Button icons should either have dividers like on the Query Builder page, or the Query Builder page icons should match this sans-dividers style. If we remove dividers, I think there needs to be more space between the icons than what is here.
screen shot 2016-03-14 at 11 51 26 am

@kdoh: I wonder about the legibility of the chart axes since they're small and pretty faded. I feel like we should show this to someone older than 45.

**Comment by tlrobinson** [comment]
Thanks @mazameli 

> I sometimes experienced a bug in Chrome where the dash wouldn't correctly scale when I went to full screen (note the huge empty white gap on the right and bottom):

Any idea what conditions trigger it? I haven't seen this behavior.

> When I hit esc to leave full Chrome's screen mode, Metabase kept the dashboard in the fullscreen style, even though the window was no longer full-screened:

This was intentional, since there's no other way to get the fullscreen layout without the browser window being fullscreen. Happy to consider alternatives though.

> Strange bug with tooltips. When hovering over the line chart to try to see the tooltips, they would weirdly show up on the chart above, like in this screenshot:

Possibly same as #2145?

> There's also a small funny gap at the top of the pie chart:

This is due to the "other" slice being very small (0.1%). We should probably hide the "other" slice if it's < ~2% (since the gap itself is 1%) but still show it in the legend?

> Many click targets that should change the cursor to a pointer are failing to do so:

I'm not sure whether things that can be hovered but not clicked should have the hand cursor, since no additional action can take place by clicking.

**Comment by agilliland** [comment]
> We should probably hide the "other" slice if it's < ~2% (since the gap itself is 1%) but still show it in the legend?

That seems really dangerous to me because then you are messing with the accuracy of the way the chart is read.  For example, if you had a pie chart that was 99% value A and 1% value B, but the pie rendered as if A was 100% then that's not good.

**Comment by tlrobinson** [comment]
> on the smallest card size (2x2) i'm only seeing 1 character of the card title with an ellipses. e.g "C..." even though there is a reasonable amount of extra space available.

I assume you're in editing mode? Maybe there's a better solutions, but currently we need to leave space for the action buttons to fade in. 

> should we continue creating new cards at the 2x2 size? that seems odd. i would expect a larger default size for new cards added to the dash.

I believe it adds them at each card type's min size, but I adjusted the values from @kdoh's designs when we were using the 12 cell grid size. @kdoh do you think the min sizes specified in the Sketch file you gave me are still accurate, or do you want to go through them and check?

**Comment by mazameli** [comment]
>> I sometimes experienced a bug in Chrome where the dash wouldn't correctly scale when I went to full screen (note the huge empty white gap on the right and bottom):

> Any idea what conditions trigger it? I haven't seen this behavior.

No, and I can't reliably reproduce it, though for a bit it was happening every time I tried to enter fullscreen mode. I'll keep an eye on it.

>> When I hit esc to leave full Chrome's screen mode, Metabase kept the dashboard in the fullscreen style, even though the window was no longer full-screened:

> This was intentional, since there's no other way to get the fullscreen layout without the browser window being fullscreen. Happy to consider alternatives though.

I'm not sure I understand. Let's chat in person?

>> Strange bug with tooltips. When hovering over the line chart to try to see the tooltips, they would weirdly show up on the chart above, like in this screenshot:

>Possibly same as #2145?

Yes, I bet it is.

>> There's also a small funny gap at the top of the pie chart:

> This is due to the "other" slice being very small (0.1%). We should probably hide the "other" slice if it's < ~2% (since the gap itself is 1%) but still show it in the legend?

I think we closed the loop on this in Slack, and I liked your suggestion of not hiding it but showing the smallest possible slice.

>> Many click targets that should change the cursor to a pointer are failing to do so:

> I'm not sure whether things that can be hovered but not clicked should have the hand cursor, since no additional action can take place by clicking.

In that case we need to make sure it's consistent, because there are some cases where we *are* making it a cursor when it's not clickable. Aside from that, I did list some icons that weren't correctly showing the hand pointer.

**Comment by kdoh** [comment]
Have a few more things after playing with the latest version apologies if there are any dupes:

- we probably don't need to display tooltips on hover for small cards when they have non truncated values like "32" for example (the logic to show 3,301 for a tooltip as 3.3k is basically applying to all values at smaller sizes i think)
- the icons on the dash header need a bit more horizontal spacing (i think @mazameli alluded to this) 
- we should have cursor pointer on table row pagination on hover
- the icons for the table row pagination icons aren't properly lined up
- i found a strange instance where I was able to paginate beyond the total number of rows
- sun and moon icons for full screen should be switched ( i should see the sun in night mode to indicate i can switch and vice versa)
- both the fullscreen and general dash heading icons need tooltips
- when you add a new card the yellow highlight fades out over time and the card background is transparent.
- it seems like when auto refresh is on it flashes the favicon?
- the card defaults are definitely a bit too small. i'm thinking we want an initial size of at least 4x4
- the horizontal table scrollbar should only show on hover
- there seems to be a bit of a delay between flipping the color on the active sorted column in tables
- the column sort icon should show when you hover on the column heading to indicate they are sortable (like the table on the query builder page)

**Comment by kdoh** [comment]
One other thing I wanted to capture so we don't forget was that after we left auto-refresh on over the weekend when I was checking the dash on the tv it made Safari beachball to the point where I had to force-quit it and couldn't switch in and out of night mode or fullscreen.

**Comment by tlrobinson** [comment]
> the icons for the table row pagination icons aren't properly lined up

@kdoh Yeah this is why I need SVG assets for those (unicode wonkiness). Can you get me those?

> there seems to be a bit of a delay between flipping the color on the active sorted column in tables

I'm not sure what this means.

**Comment by kdoh** [comment]
> @kdoh Yeah this is why I need SVG assets for those (unicode wonkiness). Can you get me those?
Derp I forgot you'd mentioned that a while back. Yup totally.

> there seems to be a bit of a delay between flipping the color on the active sorted column in tables
> I'm not sure what this means.
So if you sort by one column in a table and then the other I'm seeing it take a while for the previously sorted column heading to fade from active blue to normal again. Let me know if you can't repro it and I can try and show you.

**Comment by agilliland** [comment]
BTW, I just crossed out the issue i had listed about the single visible character.  it was indeed happening in edit mode and it makes sense that we don't have room given the need for the other action buttons.  at some point we may want to revisit that, but for now it's clearly not that big of a deal.

**Comment by camsaul** [comment]
How does this resolve #1166 (pin map zits)?

**Comment by MrMauricioLeite** [comment]
Sorry to interrupt the constructive discussion. But I've a quick comment, after testing this by myself.
Oh boy this really looks sweet.
Really, the interface makes me feel comfortable and everything works really well. Congrats guys. You are all doing an amazing job.

Will it make to 0.16.0 milestone?

**Comment by agilliland** [comment]
Yes, this will be released in 0.16.0.

**Comment by kdoh** [comment]
Thanks for the kind words @MrMauricioLeite

**Pull request merged**